packages/messaging/ios/src/RCTAEPMessaging.m

```diff
@@ -24,5 +39,127 @@ - (dispatch_queue_t)methodQueue
+        if(result) {
+            resolve(result);
+        } else {
+            reject(@"error", @"error in handleJavaScriptMessage", nil);
+        }
+    }];
+}
+
+RCT_EXPORT_METHOD(clearMessage: (NSString *) messageId) {
+    [cachedMessages removeObjectForKey:messageId];
+}
+
+RCT_EXPORT_BLOCKING_SYNCHRONOUS_METHOD(shouldShowMessage: (BOOL) shouldShowMessage) {
+    self->shouldShowMessage = shouldShowMessage;
+    NSLog(@">>>> IN JS thread %i", self->shouldShowMessage);
+    dispatch_semaphore_signal(semaphore);
+    return nil;
+}
+
+//MARK: - AEPMessagingDelegate functions.
+- (void) onDismissWithMessage:(id<AEPShowable> _Nonnull) message {
+    AEPFullscreenMessage * fullscreenMessage = (AEPFullscreenMessage *) message;
+    AEPMessage * messageObj = (AEPMessage *) fullscreenMessage.settings.parent;
+        if(messageObj) {
+            [self emitEventWithName:@"onDismiss"  body:@{@"id":messageObj.id, @"autoTrack":messageObj.autoTrack ? @"true" : @"false"}];
+        }
+}
+
+- (void) onShowWithMessage:(id<AEPShowable> _Nonnull)message {
+    AEPFullscreenMessage * fullscreenMessage = (AEPFullscreenMessage *) message;
+    AEPMessage * messageObj = (AEPMessage *) fullscreenMessage.settings.parent;
+    if(messageObj) {
+        [self emitEventWithName:@"onShow" body:@{@"id":messageObj.id, @"autoTrack":messageObj.autoTrack ? @"true" : @"false"}];
+    }
+}
+
+- (BOOL) shouldShowMessageWithMessage:(id<AEPShowable> _Nonnull)message {
+    AEPFullscreenMessage * fullscreenMessage = (AEPFullscreenMessage *) message;
+    AEPMessage * messageObj = (AEPMessage *) fullscreenMessage.settings.parent;
+    if(messageObj) {
+        NSLog(@">>>> shouldShowMessageWithMessage Message id: %@", messageObj.id);
+        [cachedMessages setObject:messageObj forKey:messageObj.id];
+        [self emitEventWithName:@"shouldShowMessage" body:@{@"id":messageObj.id, @"autoTrack":messageObj.autoTrack ? @"true" : @"false"}];
+    dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER);
+    }
+    NSLog(@">>>> From shouldShowMessageWithMessage returning %i ", self->shouldShowMessage);
+    return self->shouldShowMessage;
+}
+
+- (void) urlLoaded:(NSURL *)url byMessage:(id<AEPShowable>)message {
+    AEPFullscreenMessage * fullscreenMessage = (AEPFullscreenMessage *) message;
+    AEPMessage * messageObj = (AEPMessage *) fullscreenMessage.settings.parent;
+    if(messageObj){
+        [self emitEventWithName:@"urlLoaded" body:@{@"id":messageObj.id, @"autoTrack":messageObj.autoTrack ? @"true" : @"false"}];
+    }
+ }
+
+- (NSArray<NSString *> *) supportedEvents {
+    return @[@"onShow", @"onDismiss", @"shouldShowMessage",@"urlLoaded"];
+}
+
+- (void) startObserving {
+    hasListeners = true;
+}
+
+- (void) stopObserving {
+    hasListeners = false;
+}
+
+- (void) emitEventWithName: (NSString *) name body: (NSDictionary<NSString*, NSString*> *) dictionary {
+    if(hasListeners){
+        [self sendEventWithName:name body:dictionary];
+    }
+}
```

Comment:
is this section of functions to support the synchronous calls to and from javascript?

Reply:
This functions are used to send events(along with event data) from native to JS event listeners.
The boolean hasListeners indicates that if an event Listener exists in the JS end. Then only send the event otherwise RN will give warning.

The events are used to notify the JS about the MessageDelegate functions called. Which further in JS are used to call corresponding Message Delegate functions registered in the JS.